File: backtracking/word_search.py
from typing import List
from typing import Counter
import collections


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        visited = set()

        def exist(row: int, col: int, currWordIndex: int):
            if currWordIndex == len(word):
                return True

            if (
                min(row, col) < 0
                or row >= rows
                or col >= rows
                or board[row][col] != word[currWordIndex]
                or (row, col) in visited
            ):
                return False

            visited.add((row, col))
            res = (
                exist(row + 1, col, currWordIndex + 1)
                or exist(row - 1, col, currWordIndex + 1)
                or exist(row, col + 1, currWordIndex + 1)
                or exist(row, col - 1, currWordIndex + 1)
            )
            visited.remove((row, col))
            return res

        count = collections.defaultdict(int, sum(map(Counter, board), Counter()))
        if count[word[0]] > count[word[-1]]:
            word = word[::-1]

        for row in range(rows):
            for col in range(cols):
                if exist(col, row, 0):
                    return True

        return False


s = Solution()
print(
    s.exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "SEE")
)

File: kotlin/src/main/kotlin/neetcode_roadmap/backtracking/WordSearch.kt
package neetcode_roadmap.backtracking

fun exist(board: Array<CharArray>, word: String): Boolean {
    val visited = mutableSetOf<Pair<Int, Int>>()

    fun dfs(row: Int, col: Int, i: Int): Boolean {
        if (i == word.length) {
            return true
        }

        if (!board.indices.contains(row) || !board[row].indices.contains(col) || visited.contains(row to col) || board[row][col] != word[i]) {
            return false
        }

        visited.add(row to col)
        val res =
            dfs(row + 1, col, i + 1) || dfs(row - 1, col, i + 1) || dfs(row, col + 1, i + 1) || dfs(row, col - 1, i + 1)
        visited.remove(row to col)
        return res
    }

    for (row in board.indices) {
        for (col in board[row].indices) {
            if (dfs(row, col, 0)) {
                return true
            }
        }
    }
    return false
}

fun main() {
    val output = exist(
        board = arrayOf(
            charArrayOf('A', 'B', 'C', 'E'),
            charArrayOf('S', 'F', 'C', 'S'),
            charArrayOf('A', 'D', 'E', 'E')
        ),
        word = "ABCCED"
    )
    println(output)
}